codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java

```diff
@@ -708,7 +708,8 @@ public boolean isPushPromiseSent() {
 
         @Override
         public int incrementAndGetNextStreamId() {
-            return nextReservationStreamId >= 0 ? nextReservationStreamId += 2 : nextReservationStreamId;
+            return goAwayReceived() ? -1 :
```

Comment:
Would it be cleaner to instead set `nextReservationStreamId = -1` (or Integer.MIN_VALUE) in `goAwayReceived()`?

Reply:
Perhaps. The connection between `nextReservationStreamId` and `nextStreamIdToCreate` is somewhat subtle, but it looks safe to me. I'm happy to change it if you like.

Reply:
I wonder if we should instead change `Http2FrameCodec` from this....

```java
final int streamId = connection.local().incrementAndGetNextStreamId();
if (streamId < 0) {
  promise.setFailure(new Http2NoMoreStreamIdsException());
  return;
}
stream.id = streamId;
```

to this...

```java
final Http2Stream oldStream;
try {
  oldStream = connection.local().createStream(connection.local().incrementAndGetNextStreamId(), false);
} catch (Throwable cause) {
  promise.setFailure(cause);
  return;
}
stream.id = oldStream.id();
```

this way the `Http2FrameCodec` doesn't have to make assumptions about `Http2NoMoreStreamIdsException` or create any new failure cases.

Reply:
Right now the stream creation is driven almost entirely in the `Connection[Encoder|Decoder]`s (with the exception being h2c upgrade pathways and a stream error handling pathway for decoding headers that were too large). It makes a fair amount of sense to have it in the Encoder/Decoder due to the requirement of monotonically increasing stream-id's: having stream creation in the Encoder/Decoder helps ensure that we put HEADERS frames on the socket in the right order.

That said, it would probably work. I also don't think it's a good idea to be dolling out stream id's that look valid but aren't, but that's somewhat orthogonal other than combining both changes would result in throwing a [`Http2NoMoreStreamIdsException`](https://github.com/netty/netty/blob/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java#L873) instead of the synthetic RST_STREAM you have in https://github.com/netty/netty/pull/8069. That is arguably a good thing since, as @ejona86 noted, we don't want to trigger retries against the same session.